app/mailers/application_mailer.rb

```diff
@@ -36,4 +37,37 @@ def generate_unsubscribe_token(id, email_type)
   def use_custom_host
     ActionMailer::Base.default_url_options[:host] = Settings::General.app_domain
   end
+
+  def perform_deliveries?
+    if Rails.env.production?
+      self.perform_deliveries = Settings::SMTP.password.present? ||
+        ENV["SENDGRID_API_KEY"].present?
+    else
+      Rails.configuration.action_mailer.perform_deliveries
+    end
```

Comment:
I don't exactly like what I have here either but here's the conundrum: I would like stick settings in `development.rb` and `test.rb` hence the use of `Rails.configuration.action_mailer....`. Also, for reason beyond me, if `User#confirmation_required?` is false majority of the spec fails. 

Would love to hear what others think of this.

Reply:
**blocking**: I don't think this is going to work: 

1. `perform_deliveries?` is called in `ForemInstance.smtp_enabled?`, which shouldn'tr trigger this as a side effect
2. we should separate the dynamic behavior for delivery from the question "is smtp enabled?"

For the first step I believe we should follow the documentation's instructions:

> Using an after_action callback also enables you to override delivery method settings by updating mail.delivery_method.settings.

from https://guides.rubyonrails.org/action_mailer_basics.html#action-mailer-callbacks

So, even the modification of the `perform_deliveries` instance value should be done at runtime in a callback

Something like:

```ruby
after_action :set_perform_deliveries

def set_perform_deliveries
  self.perform_deliveries = Settings::SMTP.password.present? || ENV["SENDGRID_API_KEY"].present?
end
```

this way we don't need `if Rails.env.production?` and we can easily test it by setting each value in turn.
  
I really think we should add some tests for this logic and `set_delivery_options` :)

For the second step: what we're actually asking is if `Settings::SMTP` is configured or if the sengrid key exists. We don't actually care about `Rails.configuration.action_mailer.perform_deliveries`, do we?

We would if the settings were static, but in this case they aren't.

What if we add an `enabled?` method to `Settings::SMTP` like this?

```ruby
    class << self
      def enabled?
        (user_name.present? && password.present?) || ENV["SENDGRID_API_KEY"].present?
      end
    end
```

and then use it here as:

```ruby
after_action :set_perform_deliveries

def set_perform_deliveries
  self.perform_deliveries = Settings::SMTP.enabled?
end
```

and in `ForemInstance` as:

```ruby
  def self.smtp_enabled?
    Settings::SMTP.enabled?
  end
```

Regardless of where we are, production, testing or development, it will have the same behavior

Reply:
> What if we add an enabled? method to Settings::SMTP like this? [...] and in ForemInstance as:

I think it should be sufficient if we keep this only in `ForemInstance` and keep the settings classes themselves relatively "dumb":

```ruby
class ForemInstance
  ...
  def self.smtp_enabled?
    (Settings::SMTP.user_name.present? && Settings::SMTP.password.present?) ||
      ENV["SENDGRID_API_KEY"].present?
  end
end
```

Semantically this makes more sense to me as I want to "ask" the Forem instance if it has SMTP support enabled, where this information comes from is pretty secondary.

Reply:
Thank you both for the reviews! 

I went ahead and applied a combination of both of your suggestions. I tested locally and everything is working smoothly :+1: 

Also, based on the spec failure patterns and manually testing: `after_action` either cannot change `perform_deliveries` or it's too late to do it. Putting two email specs back to back with `after_action`; the first spec will pass but the second fails because the email isn't sent on the second. Using `before_action`, both email spec fails because neither was sent out.